### PR TITLE
Add ut for MysqlTask

### PR DESCRIPTION
**Pull request by Little-Wallace**
### What problem does this PR solve? 

Add some ut for https://github.com/pingcap/dm/pull/424

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

No

### CODE REVIEW

File: dm/worker/hub.go [hunk 1]
```diff
@@ -29,11 +29,9 @@ type ConditionHub struct {
 
 // InitConditionHub inits the singleton instance of ConditionHub
 func InitConditionHub(w *Worker) {
-	once.Do(func() {
```
Comment (csuzhangxc) [inline dm/worker/hub.go]: How about removing `once` definition too?

File: dm/master/server.go [hunk 2]
```diff
@@ -289,7 +289,7 @@ func (s *Server) RegisterWorker(ctx context.Context, req *pb.RegisterWorkerReque
 		}
 	}
 	fmt.Println("=======Register a worker")
-	s.coordinator.AddWorker(req.Name, req.Address)
+	s.coordinator.AddWorker(req.Name, req.Address, nil)
```
Comment (csuzhangxc) [inline dm/master/server.go]: Why do we pass a `nil` client here? or when do we need a non-`nil` client? (just for testing?)

File: dm/master/server.go [hunk 3]
```diff
@@ -1284,7 +1275,7 @@ func (s *Server) fetchWorkerDDLInfo(ctx context.Context) {
 				default:
 					resp, err := cli.SendRequest(ctx, request, s.cfg.RPCTimeout)
 					if err != nil {
-						log.L().Error("create FetchDDLInfo stream", zap.String("worker", worker), log.ShortError(err))
+						log.L().Error("create FetchDDLInfo stream", zap.String("worker", source), log.ShortError(err))
```
Comment (csuzhangxc) [inline dm/master/server.go]:
```suggestion
						log.L().Error("create FetchDDLInfo stream", zap.String("source", source), log.ShortError(err))
```

ditto for the following lines.

File: dm/master/server.go [hunk 4]
```diff
@@ -1257,14 +1253,9 @@ func (s *Server) fetchWorkerDDLInfo(ctx context.Context) {
...
-	for _, w := range s.coordinator.GetWorkersByStatus(coordinator.WorkerBound) {
-		client, err := w.GetClient()
-		if err != nil {
-			log.L().Error("cannot get worker client", zap.String("name", w.Name()))
-			continue
-		}
+	for source, w := range s.coordinator.GetRunningMysqlSource() {
 		wg.Add(1)
-		go func(worker string, cli workerrpc.Client) {
+		go func(source string, cli *coordinator.Worker) {
```
Comment (csuzhangxc) [inline dm/master/server.go]: How about updating the name of `cli`?

File: dm/master/server_test.go [hunk 5]
```diff
@@ -1453,3 +1459,83 @@ func (t *testMaster) TestJoinMember(c *check.C) {
...
+	cfg1.MasterAddr = tempurl.Alloc()[len("http://"):]
+	cfg1.PeerUrls = tempurl.Alloc()
+	cfg1.AdvertisePeerUrls = cfg1.PeerUrls
+	cfg1.InitialCluster = fmt.Sprintf("%s=%s", cfg1.Name, cfg1.AdvertisePeerUrls)
+
+	s1 := NewServer(cfg1)
+	c.Assert(s1.Start(ctx), check.IsNil)
+	defer s1.Close()
+	mysqlCfg := config.NewMysqlConfig()
+	mysqlCfg.LoadFromFile("./dm-mysql.toml")
```
Comment (csuzhangxc) [inline dm/master/server_test.go]: check the returned `err`.

File: dm/master/server_test.go [hunk 6]
```diff
@@ -1453,3 +1459,83 @@ func (t *testMaster) TestJoinMember(c *check.C) {
...
+		Msg:    "",
+	}, nil)
+	req.Op = pb.WorkerOp_StartWorker
+	mockWorkerClient.EXPECT().OperateMysqlTask(
+		gomock.Any(),
+		req,
+	).Return(&pb.MysqlTaskResponse{
+		Result: true,
+		Msg:    "",
+	}, nil)
```
Comment (csuzhangxc) [inline dm/master/server_test.go]: How about putting them closer to the corresponding `OperateMysqlWorker`?

File: dm/worker/server_test.go [hunk 7]
```diff
@@ -140,3 +137,39 @@ func (t *testServer) createClient(c *C, addr string) pb.WorkerClient {
 	c.Assert(err, IsNil)
 	return pb.NewWorkerClient(conn)
 }
+
+func (t *testServer) testOperateWorker(c *C, s *Server, start bool) {
+	dir := c.MkDir()
+	workerCfg := &config.MysqlConfig{}
+	workerCfg.LoadFromFile("./dm-mysql.toml")
```
Comment (csuzhangxc) [inline dm/worker/server_test.go]: check the returned `err`?

File: dm/master/server.go (hunk 2 above)
Comment (Little-Wallace) [reply dm/master/server.go]: Yes. Only for testing.

File: dm/master/server.go (hunk 3 above)
Comment (Little-Wallace) [reply dm/master/server.go]: ok.

File: dm/master/server.go (hunk 4 above)
Comment (Little-Wallace) [reply dm/master/server.go]: ok

File: dm/worker/join.go [hunk 8]
```diff
@@ -83,7 +83,7 @@ func (s *Server) KeepAlive() (bool, error) {
 	if err != nil {
 		return false, err
 	}
-	ch, err := s.etcdClient.KeepAlive(s.ctx, lease.ID)
+	ch, err := s.etcdClient.KeepAlive(cliCtx, lease.ID)
```
Comment (nolouch) [inline dm/worker/join.go]: use the global context.

File: dm/common/common.go [hunk 9]
```diff
@@ -27,6 +27,9 @@ var (
 	// WorkerKeepAliveKeyAdapter used to encode and decode keepalive key.
 	// k/v: Encode(addr,name) -> time
 	WorkerKeepAliveKeyAdapter KeyAdapter = keyHexEncoderDecoder("/dm-worker/a/")
+	// WorkerConfigKeyAdapter the config path of upstream.
+	// k/v: Encode(source-id) -> addr
+	WorkerConfigKeyAdapter KeyAdapter = keyEncoderDecoder("/dm-worker/config/")
```
Comment (csuzhangxc) [inline dm/common/common.go]: Is `Encode(addr) -> source-id`? one usage is `clientv3.OpPut(common.WorkerConfigKeyAdapter.Encode(s.cfg.WorkerAddr), cfg.SourceID)`

File: dm/master/coordinator/coordinator.go [hunk 10]
```diff
@@ -295,36 +310,122 @@ func (c *Coordinator) ObserveWorkers() {
...
+	fmt.Printf("schedule source : %v\n", source)
+	// c.waitingTask <- source
+}
+
+func (c *Coordinator) unschedule(cli *clientv3.Client, cfg config.MysqlConfig) {
+	delete(c.taskConfigs, cfg.SourceID)
+	// TODO: store taskConfigs in ETCD
+}
 
+func (c *Coordinator) tryRestartMysqlTask() {
```
Comment (csuzhangxc) [inline dm/master/coordinator/coordinator.go]: Is this not be called in any place yet?

File: dm/master/coordinator/coordinator.go [hunk 11]
```diff
@@ -295,36 +310,122 @@ func (c *Coordinator) ObserveWorkers() {
...
+	fmt.Printf("tryRestartMysqlTask")
+	hasTaskToSchedule := true
+	for hasTaskToSchedule {
+		select {
+		case source := <-c.waitingTask:
+			if cfg, ok := c.taskConfigs[source]; ok {
+				ret := false
+				if w, ok := c.upstreams[source]; ok {
+					// Try start mysql task at the same worker.
+					c.mu.RUnlock()
```
Comment (csuzhangxc) [inline dm/master/coordinator/coordinator.go]: Which one is this `RUnlock` matched `RLock`?

File: dm/master/server.go [hunk 12]
```diff
@@ -1652,89 +1645,87 @@ func (s *Server) CheckTask(ctx context.Context, req *pb.CheckTaskRequest) (*pb.C
...
+				}, nil
+			}
+		} else {
+			if resp, err = w.OperateMysqlWorker(ctx, req, s.cfg.RPCTimeout); err != nil {
+				return &pb.MysqlWorkerResponse{
+					Result: false,
+					Msg:    errors.ErrorStack(err),
+				}, nil
+			}
+			s.coordinator.HandleStoppedWorker(w, cfg)
```
Comment (csuzhangxc) [inline dm/master/server.go]: Do not need to handle the returned result here?

File: dm/worker/server.go [hunk 13]
```diff
@@ -194,6 +200,41 @@ func (s *Server) checkWorkerStart() *Worker {
...
+	}
+	if s.worker.cfg.SourceID != sourceID {
+		return nil
+	}
+	w := s.worker
+	s.worker = nil
+	return w
+}
+
+func (s *Server) retryWriteEctd(ops ...clientv3.Op) string {
```
Comment (csuzhangxc) [inline dm/worker/server.go]: How about returning `error` instead of `string`?

File: dm/common/common.go (hunk 9 above)
Comment (Little-Wallace) [reply dm/common/common.go]: Yes. It was a mistake.

File: dm/master/coordinator/coordinator.go [hunk 14]
```diff
@@ -83,13 +91,32 @@ func (c *Coordinator) Start(ctx context.Context, etcdClient *clientv3.Client) er
...
+		sourceID := common.UpstreamConfigKeyAdapter.Decode(string(kv.Key))[0]
+		cfgStr := string(kv.Value)
+		cfg := config.NewMysqlConfig()
+		err = cfg.Parse(cfgStr)
+		if err != nil {
+			log.L().Error("cannot parse config", zap.String("source", sourceID), zap.Error(err))
+			continue
+		}
+		c.taskConfigs[sourceID] = *cfg
+		c.schedule(sourceID)
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply dm/master/coordinator/coordinator.go]: These tasks  will running in gorutine `ObserveWorkers`.

File: dm/master/coordinator/coordinator.go [hunk 15]
```diff
@@ -161,68 +189,55 @@ func (c *Coordinator) HandleStartedWorker(w *Worker, cfg *config.MysqlConfig, su
 	defer c.mu.Unlock()
 	if succ {
 		c.upstreams[cfg.SourceID] = w
-		c.configs[w.Address()] = *cfg
+		c.workerToConfigs[w.Address()] = cfg.SourceID
```
[the remark this answers is not in the document]
Comment (Little-Wallace) [reply dm/master/coordinator/coordinator.go]: Emmmm. I did not thought out a better name. Do you have any idea ?

File: dm/master/coordinator/coordinator.go (hunk 11 above)
Comment (Little-Wallace) [reply dm/master/coordinator/coordinator.go]: The coodinator should unlock mutex when it calls a RPC.

File: dm/master/server.go (hunk 12 above)
Comment (Little-Wallace) [reply dm/master/server.go]: No, because it must be success

File: dm/worker/server.go (hunk 13 above)
Comment (Little-Wallace) [reply dm/worker/server.go]: Emmmm, it is just a warning message.

File: dm/master/coordinator/coordinator.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply dm/master/coordinator/coordinator.go]: `workerToSources`?

File: dm/master/server.go [hunk 16]
```diff
@@ -212,6 +212,28 @@ func (s *Server) Start(ctx context.Context) (err error) {
...
+	resp, err := s.etcdClient.Get(ectx, common.UpstreamSubTaskKeyAdapter.Path(), clientv3.WithPrefix())
+	if err != nil {
+		return err
+	}
+	for _, kv := range resp.Kvs {
+		infos := common.UpstreamSubTaskKeyAdapter.Decode(string(kv.Key))
+		sourceID := infos[0]
+		taskName := infos[1]
+		if sources, ok := s.taskSources[taskName]; ok {
+			sources = append(sources, sourceID)
```
Comment (csuzhangxc) [inline dm/master/server.go]: Is `sources` used anywhere?

**Pull request merged**